From 2e62a5d1977b14b58062214f0c4add9b50b7c364 Mon Sep 17 00:00:00 2001
From: Sai Vemuri <svemuri@redhat.com>
Date: Thu, 11 Sep 2014 22:43:06 -0400
Subject: [infiniband] cxgb4: rmb() after reading valid gen bit

Message-id: <1410475447-94392-48-git-send-email-svemuri@redhat.com>
Patchwork-id: 93796
O-Subject: [RHEL7.1 PATCH BZ 1124947 047/108] RDMA/cxgb4: rmb() after reading valid gen bit
Bugzilla: 1124947
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

Some HW platforms can reorder read operations, so we must rmb() after
we see a valid gen bit in a CQE but before we read any other fields
from the CQE.

Signed-off-by: Steve Wise <swise@opengridcomputing.com>
Signed-off-by: Roland Dreier <roland@purestorage.com>
(cherry picked from commit def4771f4bf428d39c7fe6006a9e1a20ee380d1e)
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/infiniband/hw/cxgb4/t4.h b/drivers/infiniband/hw/cxgb4/t4.h
index 931bfd1..1f329fa 100644
--- a/drivers/infiniband/hw/cxgb4/t4.h
+++ b/drivers/infiniband/hw/cxgb4/t4.h
@@ -620,6 +620,9 @@ static inline int t4_next_hw_cqe(struct t4_cq *cq, struct t4_cqe **cqe)
   printk(KERN_ERR MOD "cq overflow cqid %u\n", cq->cqid);
   BUG_ON(1);
  } else if (t4_valid_cqe(cq, &cq->queue[cq->cidx])) {
+
+  /* Ensure CQE is flushed to memory */
+  rmb();
   *cqe = &cq->queue[cq->cidx];
   ret = 0;
  } else
-- 
1.7.1